staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -424,30 +424,95 @@ func (wc *watchChan) startWatching(watchClosedCh chan struct{}, initialEventsEnd
 	close(watchClosedCh)
 }
 
-// processEvent processes events from etcd watcher and sends results to resultChan.
-func (wc *watchChan) processEvent(wg *sync.WaitGroup) {
-	defer wg.Done()
+// processEvents processes events from etcd watcher and sends results to resultChan.
+func (wc *watchChan) processEvents(wg *sync.WaitGroup) {
+	processing := concurrentOrderedEventProcessing{
+		input:       wc.incomingEventChan,
+		processFunc: wc.transform,
+		output:      wc.resultChan,
 
+		objectType:    wc.watcher.objectType,
+		groupResource: wc.watcher.groupResource,
+	}
+	processing.run(wc.ctx, wg, processEventConcurrency)
+}
+
+type concurrentOrderedEventProcessing struct {
+	input       chan *event
+	processFunc func(*event) *watch.Event
+	output      chan watch.Event
+
+	// Metadata for logging
+	objectType    string
+	groupResource schema.GroupResource
+}
+
+func (p *concurrentOrderedEventProcessing) run(ctx context.Context, wg *sync.WaitGroup, concurrency int) {
+	processingQueue := make(chan chan *watch.Event, concurrency-1)
+	wg.Add(1)
+	go func() {
```

Comment:
I don't have a clear sense for whether this was previously on the request-handling path and would propagate panics upward to the request handler which would terminate the request without crashing the server.

Each spawned goroutine that does not specifically propagate panics to its parent goroutine makes any panic within the goroutine server-crashing. Can we verify what the behavior before was and determine if we need to handle / propagate these here?

Reply:
@serathius - please look into that and address this comment

Reply:
Yes, we didn't propagate the panic to request handling. If there was a panic in the decoding it will cause the event processing goroutine to terminate and blockthe watch. This seems not very good, especially for watch opened by watch cache that can be more persistent, however this PR doesn't change the behavior I have written a test that confirms that there is no difference here. See the following code:

```
func TestWatchCodec(t *testing.T) {

	tcs := []struct{
		name string
		codec runtime.Codec
		expectTimeout bool
	} {
		{
			name: "Success",
		},
		{
			name:  "Panic",
			codec: decodePanicOnWatch{Codec: apitesting.TestCodec(codecs, examplev1.SchemeGroupVersion)},
			expectTimeout: true,
		},
	}
	for _, tc := range tcs {
		t.Run(tc.name, func(t *testing.T) {
			opts := []setupOption{}
			if tc.codec != nil {
				opts = append(opts, withCodec(tc.codec))
			}
			ctx, store, _ := testSetup(t, opts...)

			barThird := &example.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: "third", Name: "bar"}}
			podKey := fmt.Sprintf("/pods/%s/%s", barThird.Namespace, barThird.Name)
			storedObj := &example.Pod{}

			w, err := store.Watch(ctx, "/pods", storage.ListOptions{
				Predicate:            storage.Everything,
				Recursive:            true,
			})
			if err != nil {
				t.Fatal(err)
			}
			defer w.Stop()

			out := &example.Pod{}
			err = store.Create(ctx, podKey, storedObj, out,0)
			if err != nil {
				t.Fatal(err)
			}
			var timeout bool
			select {
			case <-w.ResultChan():
			case <-time.After(time.Second):
				timeout = true
			}
			if timeout != tc.expectTimeout {
				t.Errorf("Expected to timeout on waiting for watch due to panic when decoding")
			}
		})
	}
}

type decodePanicOnWatch struct {
	runtime.Codec
}

func (f decodePanicOnWatch) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
	stackTraceStr := string(debug.Stack())
	if strings.Contains(stackTraceStr, ".transform") {
		panic("just panic")
	}
	return f.Codec.Decode(data, defaults, into)
}
```